source/fundamentals/data-formats/codecs.txt

```diff
@@ -11,38 +11,337 @@ Codecs
    :class: singlecol
 
 
+.. _data-formats-codecs:
+
 Overview
 --------
 
-- Description 
-  - Relationship between Codec, CodecRegistry, and CodecProvider
+In this guide, you can learn about **Codecs** and the supporting classes that
+handle the encoding and decoding of Java objects to BSON data in MongoDB.
+The ``Codec`` abstraction allows you to map types and fields directly to
+your domain objects rather than to an intermediate map-based one such
+as ``Document`` or ``BsonDocument``.
+
+We describe the following interfaces and classes and provide examples of
+how you can specify custom encoding and decoding logic using them in the
+following sections:
+
+- :ref:`Codec <codecs-codec>`
+- :ref:`CodecRegistry <codecs-codecregistry>`
+- :ref:`CodecProvider <codecs-codecprovider>`
+- :ref:`Custom Codec Example <codecs-custom-example>`
+
+If you are customizing encoding and decoding logic for Plain old Java objects
+(POJOs), read our guide on :doc:`POJO Customization </fundamentals/data-formats/pojo-customization>`.
+
+.. _codecs-codec:
+
+Codec
+-----
+
+The ``Codec`` interface contains abstract methods for serializing and
+deserializing Java objects to BSON data. You can define your conversion logic
+between BSON and your Java object in your implementation of this interface.
+
+To implement the ``Codec`` interface, override the ``encode()``, ``decode()``,
+and ``getEncoderClass()`` abstract methods.
+
+The ``encode()`` method requires the following parameters:
+
+.. list-table::
+   :header-rows: 1
+   :widths: 10 20
+
+   * - Parameter Type
+     - Description
+
+   * - ``writer``
+     - An instance of a class that implements ``BsonWriter``, an interface type
```

Comment:
```suggestion
     - An instance of a class that implements ``BsonWriter``, and an interface type
```

Reply:
I'm not sure if you meant it the way you wrote it, but it sounded incomplete when I read this.  Applies to all instances.

Reply:
In this case, the comma is marking the start of a nonrestrictive clause to define the term "BsonWriter". The "BsonWriter" is the  "interface type that exposes methods for writingb a BSON document". If this was a list of two items, "BsonWriter" and "something else", the comma should not be included.